src/pool/srv_pool.c

```diff
@@ -3600,6 +3715,296 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade 'upgrade status' "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc);
+			D_GOTO(out_free, rc);
+		}
+		need_commit = true;
+	}
+
+	if (need_commit)
+		rc = rdb_tx_commit(tx);
+
+out_free:
+	D_FREE(hdl_uuids);
+	return rc;
+}
+
+static int ds_pool_mark_upgrade_completed(uuid_t pool_uuid,
+					  struct pool_svc *svc, int rc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			val;
+	int				rc1;
+	bool				need_put_leader = false;
+
+	if (!svc) {
+		rc1 = pool_svc_lookup_leader(pool_uuid, &svc, NULL);
+		if (rc1 != 0)
+			return rc1;
+	}
+
+	rc1 = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc1 != 0)
+		D_GOTO(out_svc, rc1);
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	if (rc == 0)
+		val = DAOS_UPGRADE_STATUS_COMPLETED;
+	else
+		val = DAOS_UPGRADE_STATUS_FAILED;
+
+	d_iov_set(&value, &val, sizeof(val));
+	rc1 = rdb_tx_update(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+			   &value);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	rc1 = rdb_tx_commit(&tx);
+out_tx:
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	if (need_put_leader)
+		pool_svc_put_leader(svc);
+
+	return rc1;
+}
+
+static int
+ds_pool_upgrade_if_needed(uuid_t pool_uuid, struct rsvc_hint *po_hint,
+			  struct pool_svc *svc, crt_rpc_t *rpc)
+{
+	struct rdb_tx			tx;
+	d_iov_t				value;
+	uint32_t			upgrade_status;
+	uint32_t			global_version;
+	int				rc, rc1;
+	bool				upgraded = false;
+	bool				need_put_leader = false;
+
+	if (!svc) {
+		rc = pool_svc_lookup_leader(pool_uuid, &svc, po_hint);
+		if (rc != 0)
+			return rc;
+		need_put_leader = true;
+	}
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	d_iov_set(&value, &global_version, sizeof(global_version));
+	rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_global_version,
+			   &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_tx, rc);
+	} else if (rc == -DER_NONEXIST) {
+		if (!need_put_leader)
+			D_GOTO(out_tx, rc = 0);
+		D_GOTO(out_upgrade, rc);
+	} else {
+		d_iov_set(&value, &upgrade_status, sizeof(upgrade_status));
+		rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+				   &value);
+		if (rc)
+			D_GOTO(out_tx, rc);
+
+		if (global_version > DAOS_PO_GLOBAL_VERSION) {
+			D_ERROR("downgrading pool is unsupported\n");
+			D_GOTO(out_tx, rc = -DER_INVAL);
+		}
+		switch (upgrade_status) {
+		case DAOS_UPGRADE_STATUS_NOT_STARTED:
+		case DAOS_UPGRADE_STATUS_COMPLETED:
+			if (global_version < DAOS_PO_GLOBAL_VERSION)
+				D_GOTO(out_upgrade, rc = 0);
+			else
+				D_GOTO(out_tx, rc = 0);
+			break;
+		case DAOS_UPGRADE_STATUS_FAILED:
+			if (global_version < DAOS_PO_GLOBAL_VERSION) {
+				D_ERROR("upgrading failed upgraded pool again unsupported\n");
+				D_GOTO(out_tx, rc = -DER_INVAL);
+			}
+			/* try again as users requested. */
+			D_GOTO(out_upgrade, rc = 0);
+		case DAOS_UPGRADE_STATUS_IN_PROGRESS:
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment